cpp/cuproj/benchmarks/fixture/benchmark_fixture.hpp

```diff
@@ -0,0 +1,91 @@
+/*
+ * Copyright (c) 2019-2023, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <benchmark/benchmark.h>
+#include <rmm/mr/device/cuda_memory_resource.hpp>
+#include <rmm/mr/device/owning_wrapper.hpp>
+#include <rmm/mr/device/per_device_resource.hpp>
+#include <rmm/mr/device/pool_memory_resource.hpp>
+
+namespace cuspatial {
+
+namespace {
+// memory resource factory helpers
+inline auto make_cuda() { return std::make_shared<rmm::mr::cuda_memory_resource>(); }
+
+inline auto make_pool()
+{
+  return rmm::mr::make_owning_wrapper<rmm::mr::pool_memory_resource>(make_cuda());
+}
+}  // namespace
+
+/**
+ * @brief Google Benchmark fixture for libcuspatial benchmarks
+ *
+ * libcuspatial benchmarks should use a fixture derived from this fixture class to
+ * ensure that the RAPIDS Memory Manager pool mode is used in benchmarks, which
+ * eliminates memory allocation / deallocation performance overhead from the
+ * benchmark.
+ *
+ * The SetUp and TearDown methods of this fixture initialize RMM into pool mode
+ * and finalize it, respectively. These methods are called automatically by
+ * Google Benchmark
+ *
+ * Example:
+ *
+ * template <class T>
+ * class my_benchmark : public cuspatial::benchmark {
+ * public:
+ *   using TypeParam = T;
+ * };
+ *
+ * Then:
+ *
+ * BENCHMARK_TEMPLATE_DEFINE_F(my_benchmark, my_test_name, int)
+ *   (::benchmark::State& state) {
+ *     for (auto _ : state) {
+ *       // benchmark stuff
+ *     }
+ * }
+ *
+ * BENCHMARK_REGISTER_F(my_benchmark, my_test_name)->Range(128, 512);
+ */
+class benchmark : public ::benchmark::Fixture {
+ public:
+  virtual void SetUp(const ::benchmark::State& state) override
```

Comment:
Just a side mention that `SetUp` is ran by every test case of this fixture. If you have some resource that's shared by every test (pool mr can be shared right?), then `static void SetUpTestSuite()` might save you more initialization time.

https://google.github.io/googletest/advanced.html#sharing-resources-between-tests-in-the-same-test-suite

Reply:
Concern was you might end up benchmarking a potentially fragmented pool.

Reply:
A second thought: this may not be a good idea because we don't one a single large pool addressing for all test suites... This could make the pool allocation quite fragmented because of frequent alloc/dealloc. Also it could make the pool grow too large.

Reply:
I also don't think pool construction and destruction is painfully expensive.